README.md
# flutter_slide_done with null safety

Credits to [xuelongqy](https://github.com/xuelongqy/flutter_slide_done)

[![License](https://img.shields.io/badge/license-MIT-green.svg)](/LICENSE)
[![Awesome Flutter](https://img.shields.io/badge/Awesome-Flutter-blue.svg?longCache=true&style=flat-square)](https://stackoverflow.com/questions/tagged/flutter?sort=votes)
[![Pub](https://img.shields.io/badge/pub-v0.0.2+1-orange.svg)](https://pub.dartlang.org/packages/slide_done_nully)

## English Version
SlideDone is inspired by sliding unlocking, which can be defined as a slide switch that can be unidirectional 
or bidirectional. You can customize the parameters of SlideDone to make it look what you want.

## Features:

 - Support for sliding trigger events
 - Support for active trigger events
 - Support for custom styles
 - Support one-way and two-way, initialize the default location

## Demo
[Download APK-Demo](https://github.com/xuelongqy/flutter_slide_done/raw/master/art/pkg/SlideDone.apk)

![](https://github.com/xuelongqy/flutter_slide_done/raw/master/art/image/apk_QRCode.png)

## Project demonstration
|Example|
|:---:|
|![](https://raw.githubusercontent.com/xuelongqy/flutter_slide_done/master/art/image/snipaste_index.png)
|[main.dart](https://github.com/xuelongqy/flutter_slide_done/blob/master/example/lib/main.dart)|

## Simple use case
#### 1.Add a dependency in pubspec.yaml
```
//github way
dependencies:
  slide_done_nully:
    git:
      url: https://github.com/ulisseshen/slide_done_nully.git
      ref: master

// package way
dependencies:
  slide_done_nully: ^0.0.2

//don´t forget to run 
flutter pub get

```
#### 2.Add SlideDone to the layout file
```dart
import 'package:slide_done_nully/slide_done.dart';
....
  GlobalKey<SlideDoneNSState> slideDoneKey = GlobalKey<SlideDoneNSState>();
....
  SlideDoneNS(
    key: slideDoneKey,
    onStart: () async {
      await Future.delayed(Duration(seconds: 2), (){});
    },
    onEnd: () async {
      await Future.delayed(Duration(seconds: 2), (){});
    },
  ),
```
#### 3.Trigger action
```dart
  // If you don't need it, you don't need to set the key for SlideDone.
  slideDoneKey.currentState.callOnStart();
  slideDoneKey.currentState.callOnEnd();
```

#### 4.Add custom icons
```dart
  SlideDoneNS(
    key: slideDoneKey,
    startIcon: Icon(Icons.keyboard_arrow_right),
    endIcon:Icon(Icons.keyboard_arrow_left),
    startedIcon: Icon(Icons.done),
    endedIcon:  Icon(Icons.done)
  ),
```

# Donation
If you like my project, please in the upper right corner of the project "Star". Your support is my biggest encouragement! ^_^
You can also scan the qr code below or [![Donate to this project using Paypal](https://img.shields.io/badge/paypal-donate-yellow.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=334PPRBZTY3J8&source=url), donation to Author.

![](https://raw.githubusercontent.com/xuelongqy/donation/master/pay_alipay.jpg?raw=true) ![](https://raw.githubusercontent.com/xuelongqy/donation/master/pay_wxpay.jpg?raw=true) ![](https://raw.githubusercontent.com/xuelongqy/donation/master/pay_tencent.jpg?raw=true)

If in donation message note name, will be record to the list if you are making open source authors, donation can leave when making project address or personal home page, a link will be added to the list have the effect of promoting each other
(https://github.com/xuelongqy/donation/blob/master/DONATIONLIST.md)

### QQ Group - 554981921
#### Into the group of instructions
The group is not only solve the problem of SlideDone, any Flutter related issues can be discussed.
Just as its name, craigslist, as long as there is time, group of Lord will help you solve problems together.

## Open source licenses
 
```
 
MIT License

Copyright (c) 2018 xuelongqy

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

 
 ```


CHANGELOG.md
## [0.0.2+1] - 2022-02-13

* Update Examples
* Removed Unnecessary files

## [0.0.2] - 2022-02-13

* Migrate to null safety.
* Update SDK constraints to >=2.12.0 <3.0.0 i pubsple.yaml

## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example_slide_done/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/generated/i18n.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

// ignore_for_file: non_constant_identifier_names
// ignore_for_file: camel_case_types
// ignore_for_file: prefer_single_quotes

//This file is automatically generated. DO NOT EDIT, all your changes would be lost.
class S implements WidgetsLocalizations {
  const S();

  static const GeneratedLocalizationsDelegate delegate =
    GeneratedLocalizationsDelegate();

  static S? of(BuildContext context) => Localizations.of<S>(context, S);

  @override
  TextDirection get textDirection => TextDirection.ltr;

}

class $en extends S {
  const $en();
}

class GeneratedLocalizationsDelegate extends LocalizationsDelegate<S> {
  const GeneratedLocalizationsDelegate();

  List<Locale> get supportedLocales {
    return const <Locale>[
      Locale("en", ""),
    ];
  }

  Locale Function(List<Locale> locales, Iterable<Locale> supported) listResolution({required Locale fallback}) {
    return (List<Locale> locales, Iterable<Locale> supported) {
      if (locales == null || locales.isEmpty) {
        return fallback ?? supported.first;
      } else {
        return _resolve(locales.first, fallback, supported);
      }
    };
  }

  Locale Function(Locale locale, Iterable<Locale> supported) resolution({required Locale fallback}) {
    return (Locale locale, Iterable<Locale> supported) {
      return _resolve(locale, fallback, supported);
    };
  }

  Locale _resolve(Locale locale, Locale fallback, Iterable<Locale> supported) {
    if (locale == null || !isSupported(locale)) {
      return fallback ?? supported.first;
    }

    final Locale languageLocale = Locale(locale.languageCode, "");
    if (supported.contains(locale)) {
      return locale;
    } else if (supported.contains(languageLocale)) {
      return languageLocale;
    } else {
      final Locale fallbackLocale = fallback ?? supported.first;
      return fallbackLocale;
    }
  }

  @override
  Future<S> load(Locale locale) {
    final String? lang = getLang(locale);
    if (lang != null) {
      switch (lang) {
        case "en":
          return SynchronousFuture<S>(const $en());
        default:
          // NO-OP.
      }
    }
    return SynchronousFuture<S>(const S());
  }

  @override
  bool isSupported(Locale locale) =>
    locale != null && supportedLocales.contains(locale);

  @override
  bool shouldReload(GeneratedLocalizationsDelegate old) => false;
}

String? getLang(Locale? l) => l == null
  ? null
  : l.countryCode != null && l.countryCode!.isEmpty
    ? l.languageCode
    : l.toString();


lib/src/slide_done_null_safety.dart
import 'package:flutter/material.dart';

// 回调类型
typedef Future OnStart();
typedef Future OnEnd();

/// 滑动状态
enum SlideStatus {
  START,
  END
}

/// 滑动完成UI组件
class SlideDoneNS extends StatefulWidget {

  final double width;

  final double height;

  final ShapeBorder? shape;

  final ShapeBorder? buttonShape;

  final double elevation;
  final double buttonElevation;

  final Color backgroundColor;
  final Color startColor;
  final Color endColor;

  final Widget? startIcon;
  final Widget? endIcon;
  final Widget? startedIcon;
  final Widget? endedIcon;

  final Widget? startLoadWidget;
  final Widget? endLoadWidget;

  final Widget startText;
  final Widget startingText;
  final Widget startedText;

  final Widget endText;
  final Widget endingText;
  final Widget endedText;

  final OnStart? onStart;
  final OnEnd? onEnd;

  final SlideStatus status;
  final double padding;

  final Widget? startFillView;
  final Widget? endFillView;

  final int startedDelay;
  final int endedDelay;


  SlideDoneNS({
    Key? key,
    this.width = 250.0,
    this.height = 60.0,
    this.shape,
    this.buttonShape,
    this.elevation = 3.0,
    this.buttonElevation = 2.0,
    this.backgroundColor = const Color.fromRGBO(255, 255, 255, 1.0),
    this.startColor = Colors.blue,
    this.endColor = Colors.red,
    this.startIcon,
    this.endIcon,
    this.startedIcon,
    this.endedIcon,
    this.startLoadWidget,
    this.endLoadWidget,
    this.padding = 3.0,
    this.startFillView,
    this.endFillView,
    this.startedDelay = 1000,
    this.endedDelay = 1000,
    this.startText = const Text("Right slide on",
      style: TextStyle(
          fontSize: 20.0
      ),
    ),
    this.startingText = const Text("Turn on...",
      style: TextStyle(
          fontSize: 20.0
      ),
    ),
    this.startedText = const Text("It's done",
      style: TextStyle(
          fontSize: 20.0
      ),
    ),
    this.endText = const Text("Left slide off",
      style: TextStyle(
          fontSize: 20.0
      ),
    ),
    this.endingText = const Text("Turn off...",
      style: TextStyle(
          fontSize: 20.0
      ),
    ),
    this.endedText = const Text("It's done",
      style: TextStyle(
          fontSize: 20.0
      ),
    ),
    this.onStart,
    this.onEnd,
    this.status = SlideStatus.START,
  }): super(key: key);

  @override
  SlideDoneNSState createState() => SlideDoneNSState();
}
class SlideDoneNSState extends State<SlideDoneNS> with TickerProviderStateMixin<SlideDoneNS>{
  // 状态
  late SlideStatus _status;
  // 滑动范围
  late double _slideRange;
  // 滑动距离
  late double _slideDistance;
  // 上一个点
  late double _lastOffsetX;
  // 是否怎在加载
  bool _isLoad = false;
  // 是否需要改变状态(用于动画过程中判断)
  bool _isChangeStatus = false;
  // 是否加载结束等待中
  bool _isLoadedWaiting = false;
  // 动画控制
  late Animation<double> _animation;
  late AnimationController _animationController;

  // 初始化
  @override
  void initState() {
    super.initState();
    _status = widget.status;
    // 初始化范围和距离
    _slideRange = widget.width - widget.height;
    _slideDistance = widget.status == SlideStatus.START ? 0.0 : _slideRange;
    // 初始化滑动归位动画
    _animationController = new AnimationController(duration: const Duration(milliseconds: 250), vsync: this);
    _animation = new Tween(begin: 0.0, end: _slideRange).animate(_animationController)
      ..addListener(() {
        // 判断是左滑还是右滑
        if (_status == SlideStatus.START && _isChangeStatus || _status == SlideStatus.END && !_isChangeStatus) {
          _onSlide(_animation.value);
        }else {
          _onSlide(-_animation.value);
        }
      });
    _animation.addStatusListener((status) {
      if (status == AnimationStatus.forward) {
        // 初始化偏移量
        _lastOffsetX = 0.0;
        // 计算是否需要改变状态
        if (_isChangeStatus) return;
        if (_status == SlideStatus.START && _slideDistance > _slideRange - widget.height + widget.padding) {
          _isChangeStatus = true;
        }else if (_status == SlideStatus.END && _slideDistance < widget.height - widget.padding) {
          _isChangeStatus = true;
        }else {
          _isChangeStatus = false;
        }
      }else if (status == AnimationStatus.completed) {
        // 初始化偏移量
        _lastOffsetX = 0.0;
        _animationController.reset();
        // 是否需要触发事件
        if (!_isChangeStatus) return;
        if (_status == SlideStatus.START) {
          // 判断是否滑动到最右端
          _onEnd();
        } else {
          // 判断是否滑动到最左端
          _onStart();
        }
        _isChangeStatus = false;
      }
    });
  }

  // 滑动
  void _onSlide(double offsetX) {
    setState(() {
      _slideDistance += (offsetX - _lastOffsetX);
      if (_slideDistance < 0.0) {
        _slideDistance = 0.0;
      }else if (_slideDistance > _slideRange) {
        _slideDistance = _slideRange;
      }
    });
    _lastOffsetX = offsetX;
  }

  // 触发开始
  void callOnStart() {
    if (_isLoad) return;
    if (_animationController.isAnimating) return;
    if (_status == SlideStatus.START) return;
    _isChangeStatus = true;
    _animationController.forward();
  }

  // 触发结束
  void callOnEnd() {
    if (_isLoad) return;
    if (_animationController.isAnimating) return;
    if (_status == SlideStatus.END) return;
    _isChangeStatus = true;
    _animationController.forward();
    print("sss");
  }

  // 开始
  void _onStart() async {
    if (widget.onStart != null) {
      setState(() {
        _status = SlideStatus.START;
        _isLoad = true;
      });
      await widget.onStart!();
      setState(() {
        _isLoad = false;
        _isLoadedWaiting = true;
      });
      // 完成等待延时
      Future.delayed(Duration(milliseconds: widget.startedDelay), () {
        setState(() {
          _isLoadedWaiting = false;
        });
      });
    }else {
      setState(() {
        _status = SlideStatus.START;
        _isLoad = false;
        _isLoadedWaiting = false;
      });
    }
  }

  // 结束
  void _onEnd() async {
    if (widget.onEnd != null) {
      setState(() {
        _status = SlideStatus.END;
        _isLoad = true;
      });
      await widget.onEnd!();
      _isLoad = false;
      setState(() {
        _isLoad = false;
        _isLoadedWaiting = true;
      });
      // 完成等待延时
      Future.delayed(Duration(milliseconds: widget.endedDelay), () {
        setState(() {
          _isLoadedWaiting = false;
        });
      });
    }else {
      setState(() {
        _status = SlideStatus.END;
        _isLoad = false;
        _isLoadedWaiting = false;
      });
    }
  }

  // 启动动画
  void _startAnimation() {
    if (!_animationController.isAnimating) {
      _animationController.forward();
    }
  }

  @override
  Widget build(BuildContext context) {
    // 判断按钮
    Widget icon = _getIcon();
    // 按钮颜色
    Color iconColor = _getIconColor();
    // 开始文字
    Widget startText = _getStartText();
    // 结束文字
    Widget endText = _getEndText();
    return new Card(
      clipBehavior: Clip.hardEdge,
      color: widget.backgroundColor,
      elevation: widget.elevation,
      shape: widget.shape ?? RoundedRectangleBorder(
          borderRadius: BorderRadius.all(Radius.circular(widget.height / 2))
      ),
      child: Container(
        padding: EdgeInsets.all(widget.padding),
        width: widget.width,
        height: widget.height,
        child: Stack(
          children: <Widget>[
            // 结束文字
            new Container(
                width: widget.width - widget.padding * 2,
                height: widget.height - widget.padding * 2,
                child: Container(
                  margin: EdgeInsets.only(right: widget.height - widget.padding * 2 + (_slideRange - _slideDistance)),
                  child: Stack(
                    children: <Widget>[
                      Positioned(
                        child: Container(
                          width: widget.width - widget.padding - widget.height,
                          height: widget.height - widget.padding * 2,
                          child: Center(
                            child: endText,
                          ),
                        ),
                        right: -(_slideRange - _slideDistance),
                      )
                    ],
                  ),
                )
            ),
            // 开始文字
            new Container(
                width: widget.width - widget.padding * 2,
                height: widget.height - widget.padding * 2,
                child: Container(
                  margin: EdgeInsets.only(left: widget.height - widget.padding * 2 + _slideDistance),
                  child: Stack(
                    children: <Widget>[
                      Positioned(
                        child: Container(
                          width: widget.width - widget.padding - widget.height,
                          height: widget.height - widget.padding * 2,
                          child: Center(
                            child: startText,
                          ),
                        ),
                        left: -_slideDistance,
                      )
                    ],
                  ),
                )
            ),
            // 滑动按钮
            new Container(
              margin: EdgeInsets.only(left: _slideDistance),
              width: widget.height - widget.padding * 2,
              height: widget.height - widget.padding * 2,
              child: Card(
                  clipBehavior: Clip.hardEdge,
                  elevation: widget.buttonElevation,
                  margin: EdgeInsets.all(0.0),
                  color: iconColor,
                  shape: widget.buttonShape ?? RoundedRectangleBorder(
                      borderRadius: BorderRadius.all(Radius.circular((widget.height - widget.padding * 2) / 2))
                  ),
                  child: GestureDetector(
                    onHorizontalDragStart: (details){
                      if (_isLoad || _animationController.isAnimating) return;
                      _lastOffsetX = details.globalPosition.dx;
                    },
                    onHorizontalDragUpdate: (details){
                      if (_isLoad || _animationController.isAnimating) return;
                      _onSlide(details.globalPosition.dx);
                    },
                    onHorizontalDragEnd: (details){
                      if (_isLoad || _animationController.isAnimating) return;
                      // 滑动结束,启动动画
                      _startAnimation();
                    },
                    // 为按钮添加水波纹
                    child: InkWell(
                      onTap: (){},
                      borderRadius: BorderRadius.all(Radius.circular((widget.height - widget.padding * 2) / 2)),
                      child: Center(
                        child: icon,
                      ),
                    ),
                  )
              ),
            ),
          ],
        ),
      ),
    );
  }

  // 获取图标
  Widget _getIcon() {
    if (_status == SlideStatus.START) {
      if (_isLoad) {
        return widget.startLoadWidget ?? SizedBox(
            width: 25.0,
            height: 25.0,
            child: CircularProgressIndicator(
              strokeWidth: 3.0,
              valueColor: AlwaysStoppedAnimation(Colors.white),
            )
        );
      }else if (_isLoadedWaiting) {
        return widget.startedIcon ?? Icon(Icons.done,
          size: (widget.height - widget.padding * 2) / 3 * 2,
          color: Colors.white,
        );
      }else {
        return widget.startIcon ?? Icon(Icons.keyboard_arrow_right,
          size: (widget.height - widget.padding * 2) / 3 * 2,
          color: Colors.white,
        );
      }
    }else {
      if (_isLoad) {
        return widget.endLoadWidget ?? SizedBox(
            width: 25.0,
            height: 25.0,
            child: CircularProgressIndicator(
              strokeWidth: 3.0,
              valueColor: AlwaysStoppedAnimation(Colors.white),
            )
        );
      }else if (_isLoadedWaiting) {
        return widget.endedIcon ?? Icon(Icons.done,
          size: (widget.height - widget.padding * 2) / 3 * 2,
          color: Colors.white,
        );
      }else {
        return widget.endIcon ?? Icon(Icons.keyboard_arrow_left,
          size: (widget.height - widget.padding * 2) / 3 * 2,
          color: Colors.white,
        );
      }
    }
  }

  // 获取开始文字
  Widget _getStartText() {
    if (_status == SlideStatus.START) {
      if (_isLoad) {
        return widget.endingText;
      }else if (_isLoadedWaiting) {
        return widget.endedText;
      }else {
        return widget.startText;
      }
    }else {
      return Container(
        child: widget.endFillView ?? Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Icon(Icons.keyboard_arrow_left),
            Icon(Icons.keyboard_arrow_left),
            Icon(Icons.keyboard_arrow_left),
            Icon(Icons.keyboard_arrow_left)
          ],
        ),
      );
    }
  }

  // 获取结束文字
  Widget _getEndText() {
    if (_status == SlideStatus.START) {
      return widget.startFillView ?? Container(
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Icon(Icons.keyboard_arrow_right),
            Icon(Icons.keyboard_arrow_right),
            Icon(Icons.keyboard_arrow_right),
            Icon(Icons.keyboard_arrow_right)
          ],
        ),
      );
    }else {
      if (_isLoad) {
        return widget.startingText;
      }else if (_isLoadedWaiting) {
        return widget.startedText;
      }else {
        return widget.endText;
      }
    }
  }

  // 获取按钮颜色
  Color _getIconColor() {
    if (_status == SlideStatus.START) {
      if (_slideDistance >= _slideRange) {
        return widget.endColor;
      }else {
        // 计算比例
        double scale = (_slideRange - _slideDistance) / (widget.height - widget.padding) * 2;
        if (scale > 1) {
          return widget.startColor;
        }else {
          return Color.fromRGBO(
              widget.startColor.red + ((1 - scale) * (widget.endColor.red - widget.startColor.red)).floor(),
              widget.startColor.green + ((1 - scale) * (widget.endColor.green - widget.startColor.green)).floor(),
              widget.startColor.blue + ((1 - scale) * (widget.endColor.blue - widget.startColor.blue)).floor(),
              widget.startColor.opacity + (1 - scale) * (widget.endColor.opacity - widget.startColor.opacity)
          );
        }
      }
    }else {
      if (_slideDistance <= 0.0) {
        return widget.startColor;
      }else {
        // 计算比例
        double scale = (_slideDistance) / (widget.height - widget.padding) * 2;
        if (scale > 1) {
          return widget.endColor;
        }else {
          return Color.fromRGBO(
              widget.endColor.red + ((1 - scale) * (widget.startColor.red - widget.endColor.red)).floor(),
              widget.endColor.green + ((1 - scale) * (widget.startColor.green - widget.endColor.green)).floor(),
              widget.endColor.blue + ((1 - scale) * (widget.startColor.blue - widget.endColor.blue)).floor(),
              widget.endColor.opacity + (1 - scale) * (widget.startColor.opacity - widget.endColor.opacity)
          );
        }
      }
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:slide_done_nully/slide_done.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Slide done demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Slide done demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  // SlideDone key
  GlobalKey<SlideDoneNSState> slideDoneKey = new GlobalKey<SlideDoneNSState>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: ListView(
          children: <Widget>[
            _basicStyle(),
            _justOpenStyle(),
            _flatStyle(),
            _customize(),
            _more()
          ],
        )// This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  // Basic
  Widget _basicStyle() {
    return Card(
        child: Container(
            padding: EdgeInsets.all(10.0),
            child: Column(
              children: <Widget>[
                Container(
                  margin: EdgeInsets.only(bottom: 5.0, top: 5.0),
                  alignment: Alignment.centerLeft,
                  child: Text("Basic",
                    style: TextStyle(fontSize: 20.0),
                  ),
                ),
                Container(
                  margin: EdgeInsets.only(bottom: 10.0),
                  width: double.infinity,
                  height: 0.5,
                  color: Colors.grey,
                ),
                SlideDoneNS(
                  key: slideDoneKey,
                  onStart: () async {
                    await Future.delayed(Duration(seconds: 2), (){});
                  },
                  onEnd: () async {
                    await Future.delayed(Duration(seconds: 2), (){});
                  },
                ),
                Container(
                  margin: EdgeInsets.only(bottom: 10.0),
                  width: double.infinity,
                  height: 10.0,
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.end,
                  children: <Widget>[
                    FlatButton(
                        onPressed: (){
                          slideDoneKey.currentState!.callOnStart();
                        },
                        child: Text("Start")
                    ),
                    FlatButton(
                        onPressed: (){
                          slideDoneKey.currentState!.callOnEnd();
                        },
                        child: Text("End")
                    ),
                  ],
                )
              ],
            )
        )
    );
  }

  // Just Open
  Widget _justOpenStyle() {
    return Card(
        child: Container(
            padding: EdgeInsets.all(10.0),
            child: Column(
              children: <Widget>[
                Container(
                  margin: EdgeInsets.only(bottom: 5.0, top: 5.0),
                  alignment: Alignment.centerLeft,
                  child: Text("Just Open",
                    style: TextStyle(fontSize: 20.0),
                  ),
                ),
                Container(
                  margin: EdgeInsets.only(bottom: 10.0),
                  width: double.infinity,
                  height: 0.5,
                  color: Colors.grey,
                ),
                SlideDoneNS(
                  endText: Text("It's done",
                    style: TextStyle(
                        fontSize: 20.0
                    ),
                  ),
                  endedIcon: Icon(Icons.done,
                    color: Colors.white,
                  ),
                  endedText: Container(),
                  startedDelay: 0,
                  onEnd: () async {
                    await Future.delayed(Duration(seconds: 2), (){});
                  },
                  onStart:null,
                ),
              ],
            )
        )
    );
  }

  // Flat style
  Widget _flatStyle() {
    return Card(
        child: Container(
            padding: EdgeInsets.all(10.0),
            child: Column(
              children: <Widget>[
                Container(
                  margin: EdgeInsets.only(bottom: 5.0, top: 5.0),
                  alignment: Alignment.centerLeft,
                  child: Text("Flat style",
                    style: TextStyle(fontSize: 20.0),
                  ),
                ),
                Container(
                  margin: EdgeInsets.only(bottom: 10.0),
                  width: double.infinity,
                  height: 0.5,
                  color: Colors.grey,
                ),
                SlideDoneNS(
                  elevation: 0.0,
                  buttonElevation: 0.0,
                  backgroundColor: Colors.black26,
                  onStart: () async {
                    await Future.delayed(Duration(seconds: 2), (){});
                  },
                  onEnd: () async {
                    await Future.delayed(Duration(seconds: 2), (){});
                  },
                ),
              ],
            )
        )
    );
  }

  // 自定义
  Widget _customize() {
    return Card(
        child: Container(
            padding: EdgeInsets.all(10.0),
            child: Column(
              children: <Widget>[
                Container(
                  margin: EdgeInsets.only(bottom: 5.0, top: 5.0),
                  alignment: Alignment.centerLeft,
                  child: Text("Customize",
                    style: TextStyle(fontSize: 20.0),
                  ),
                ),
                Container(
                  margin: EdgeInsets.only(bottom: 10.0),
                  width: double.infinity,
                  height: 0.5,
                  color: Colors.grey,
                ),
                SlideDoneNS(
                  width: 250.0,
                  height: 60.0,
                  padding: 3.0,
                  shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.all(Radius.circular(20.0))
                  ),
                  buttonShape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.all(Radius.circular(20.0 - 3.0))
                  ),
                  startIcon: Icon(Icons.lock_outline,
                    size: 30.0,
                    color: Colors.white,
                  ),
                  endIcon: Icon(Icons.lock_open,
                    size: 30.0,
                    color: Colors.white,
                  ),
                  startText: Text("右滑开锁",
                    style: TextStyle(
                        fontSize: 20.0
                    ),
                  ),
                  startingText: Text("开启中...",
                    style: TextStyle(
                        fontSize: 20.0
                    ),
                  ),
                  startedText: Text("成功开锁",
                    style: TextStyle(
                        fontSize: 20.0
                    ),
                  ),
                  endText: Text("左滑锁定",
                    style: TextStyle(
                        fontSize: 20.0
                    ),
                  ),
                  endingText: Text("锁定中...",
                    style: TextStyle(
                        fontSize: 20.0
                    ),
                  ),
                  endedText: Text("锁定完成",
                    style: TextStyle(
                        fontSize: 20.0
                    ),
                  ),
                  startFillView: Container(
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        Icon(Icons.lens, size: 10.0,),
                        Container(width: 10.0,),
                        Icon(Icons.lens, size: 10.0,),
                        Container(width: 10.0,),
                        Icon(Icons.lens, size: 10.0,),
                        Container(width: 10.0,),
                        Icon(Icons.lens, size: 10.0,),
                      ],
                    ),
                  ),
                  endFillView: Container(
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: <Widget>[
                        Icon(Icons.lens, size: 10.0,),
                        Container(width: 10.0,),
                        Icon(Icons.lens, size: 10.0,),
                        Container(width: 10.0,),
                        Icon(Icons.lens, size: 10.0,),
                        Container(width: 10.0,),
                        Icon(Icons.lens, size: 10.0,),
                      ],
                    ),
                  ),
                  onStart: () async {
                    await Future.delayed(Duration(seconds: 2), (){});
                  },
                  onEnd: () async {
                    await Future.delayed(Duration(seconds: 2), (){});
                  },
                ),
              ],
            )
        )
    );
  }

  // More
  Widget _more() {
    return Card(
        child: Container(
            padding: EdgeInsets.all(10.0),
            child: Column(
              children: <Widget>[
                Container(
                  margin: EdgeInsets.only(bottom: 5.0, top: 5.0),
                  alignment: Alignment.centerLeft,
                  child: Text("More",
                    style: TextStyle(fontSize: 20.0),
                  ),
                ),
                Container(
                  margin: EdgeInsets.only(bottom: 10.0),
                  width: double.infinity,
                  height: 0.5,
                  color: Colors.grey,
                ),
                SlideDoneNS(
                  startText: const Text("Get more",
                    style: TextStyle(
                        fontSize: 20.0
                    ),
                  ),
                  endText: const Text("QQ群:554981921",
                    style: TextStyle(
                        fontSize: 16.0
                    ),
                  ),
                  endedIcon: const Icon(Icons.done,
                    color: Colors.white,
                  ),
                  endedText: Container(),
                  startedDelay: 0,
                  endedDelay: 0,
                  onEnd: () async {
                    // await launch("mqqopensdkapi://bizAgent/qm/qr?url=http%3A%2F%2Fqm.qq.com%2Fcgi-bin%2Fqm%2Fqr%3Ffrom%3Dapp%26p%3Dandroid%26k%3DMNLtkvnn4n28UIB0gEgm2-WBmqmGWk0Q");
                  },
                  onStart: null,
                ),
              ],
            )
        )
    );
  }
}
